File: 351002/Stepanenko/src/main/java/by/bsuir/distcomp/service/MarkerService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.MarkerRequestTo;
import by.bsuir.distcomp.dto.response.MarkerResponseTo;
import by.bsuir.distcomp.entity.Marker;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.MarkerMapper;
import by.bsuir.distcomp.repository.impl.InMemoryMarkerRepository;
import by.bsuir.distcomp.repository.impl.InMemoryTweetRepository;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class MarkerService {
    private final InMemoryMarkerRepository markerRepository;
    private final InMemoryTweetRepository tweetRepository;
    private final MarkerMapper markerMapper;

    public MarkerService(InMemoryMarkerRepository markerRepository,
                        InMemoryTweetRepository tweetRepository,
                        MarkerMapper markerMapper) {
        this.markerRepository = markerRepository;
        this.tweetRepository = tweetRepository;
        this.markerMapper = markerMapper;
    }

    public MarkerResponseTo create(MarkerRequestTo dto) {
        Marker entity = markerMapper.toEntity(dto);
        Marker saved = markerRepository.save(entity);
        return markerMapper.toResponseDto(saved);
    }

    public MarkerResponseTo getById(Long id) {
        Marker entity = markerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Marker with id " + id + " not found", 40412));
        return markerMapper.toResponseDto(entity);
    }

    public List<MarkerResponseTo> getAll() {
        return markerRepository.findAll().stream()
                .map(markerMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public MarkerResponseTo update(MarkerRequestTo dto) {
        Marker existingMarker = markerRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Marker with id " + dto.getId() + " not found", 40413));
        
        Marker entity = markerMapper.toEntity(dto);
        Marker updated = markerRepository.update(entity);
        return markerMapper.toResponseDto(updated);
    }

    public void deleteById(Long id) {
        if (!markerRepository.existsById(id)) {
            throw new ResourceNotFoundException("Marker with id " + id + " not found", 40414);
        }
        markerRepository.deleteById(id);
    }

    public List<MarkerResponseTo> getMarkersByTweetId(Long tweetId) {
        if (!tweetRepository.existsById(tweetId)) {
            throw new ResourceNotFoundException("Tweet with id " + tweetId + " not found", 40415);
        }
        
        Map<Long, Set<Long>> tweetToMarkers = new HashMap<>();
        tweetRepository.findAll().forEach(tweet -> {
            if (tweet.getMarkerIds() != null) {
                tweetToMarkers.put(tweet.getId(), tweet.getMarkerIds());
            }
        });
        
        return markerRepository.findByTweetId(tweetId, tweetToMarkers).stream()
                .map(markerMapper::toResponseDto)
                .collect(Collectors.toList());
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/MarkerService.kt
package com.distcomp.service

import com.distcomp.dto.marker.MarkerRequestTo
import com.distcomp.dto.marker.MarkerResponseTo
import com.distcomp.entity.Marker
import com.distcomp.exception.MarkerNotFoundException
import com.distcomp.mapper.MarkerMapper
import com.distcomp.repository.CrudRepository
import org.springframework.stereotype.Service

@Service
class MarkerService(
    val markerMapper: MarkerMapper,
    val markerRepository: CrudRepository<Marker>
) {
    fun createMarker(markerRequestTo: MarkerRequestTo): MarkerResponseTo {
        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    fun readMarkerById(id: Long): MarkerResponseTo {
        val marker = markerRepository.findById(id)
            ?: throw MarkerNotFoundException("Marker with id $id not found")
        return markerMapper.toMarkerResponse(marker)
    }

    fun readAll(): List<MarkerResponseTo> {
        return markerRepository.findAll().map { markerMapper.toMarkerResponse(it) }
    }

    fun updateMarker(markerRequestTo: MarkerRequestTo, markerId: Long?): MarkerResponseTo {
        if (markerId == null || markerRepository.findById(markerId) == null) {
            throw MarkerNotFoundException("Marker with id $markerId not found")
        }

        val marker = markerMapper.toMarkerEntity(markerRequestTo)
        marker.id = markerId

        markerRepository.save(marker)
        return markerMapper.toMarkerResponse(marker)
    }

    fun removeMarkerById(id: Long) {
        if (markerRepository.findById(id) == null) {
            throw MarkerNotFoundException("Marker with id $id not found")
        }
        markerRepository.removeById(id)
    }
}